### PR TITLE
WIP Reconnect peer on disconnect or error

### PR DESCRIPTION
**Pull request by carnesen**
If there's a Peer error or disconnect event, better to reconnect than fail.

### CODE REVIEW

**Comment by maraoz** [comment]
IDK why travis passed (will investigate), but tests fail on this branch.

**Comment by carnesen** [comment]
Indeed they do. My fault. Will fix. No idea why Travis didn't catch it.

**Comment by carnesen** [comment]
Dang it messed up the rebase again! Sorry about that. Let me try again on a new branch.

**Comment by carnesen** [comment]
I'm deleting this messed up branch now and resubmitting on a clean one.
